README.md
# Stopwatch

[![Maven Central](https://img.shields.io/maven-central/v/org.silentsoft/stopwatch)](https://search.maven.org/artifact/org.silentsoft/stopwatch)
[![Build Status](https://app.travis-ci.com/silentsoft/stopwatch.svg?branch=main)](https://app.travis-ci.com/silentsoft/stopwatch)
[![Quality Gate Status](https://sonarcloud.io/api/project_badges/measure?project=silentsoft_stopwatch&metric=alert_status)](https://sonarcloud.io/dashboard?id=silentsoft_stopwatch)
[![Coverage](https://sonarcloud.io/api/project_badges/measure?project=silentsoft_stopwatch&metric=coverage)](https://sonarcloud.io/dashboard?id=silentsoft_stopwatch)
[![Hits](https://hits.sh/github.com/silentsoft/stopwatch.svg)](https://hits.sh)

`Stopwatch` for get rid of chore in your code.

## Usage
```java
Stopwatch stopwatch = new Stopwatch();

stopwatch.start("initialization");
// ...
stopwatch.stop();

stopwatch.start("processing");
// ...
stopwatch.stop();

stopwatch.start("rendering");
// ...
stopwatch.stop();

stopwatch.print();
```

## Result
```
|           name |     % |      ms |      s |
|----------------|-------|---------|--------|
| initialization | 59.5% | 1,234ms | 1.234s |
|     processing | 40.0% |   830ms | 0.830s |
|      rendering |  0.5% |    10ms | 0.010s |
|                |       |         |        |
|          total |  100% | 2,074ms | 2.074s |
```

## Maven Central
```xml
<dependency>
    <groupId>org.silentsoft</groupId>
    <artifactId>stopwatch</artifactId>
    <version>2.2.1</version>
</dependency>
```

## Contributing
Pull requests are welcome. For major changes, please open an issue first to discuss what you would like to change.

Please note we have a [CODE_OF_CONDUCT](https://github.com/silentsoft/stopwatch/blob/main/CODE_OF_CONDUCT.md), please follow it in all your interactions with the project.

## License
Please refer to [LICENSE](https://github.com/silentsoft/stopwatch/blob/main/LICENSE.txt).


CHANGELOG.md
# 2.2.1 (22 Jan 2022)

## Enhancements
- Add JavaDoc comments
- Minor code cleanup

# 2.2.0 (7 Sep 2021)

## New Features
- Add `start(String, Runnable)` feature to measure `Runnable` code snippets on the fly.

# 2.1.0 (8 Aug 2021)

## New Features
- Add `pause()` and `resume()` features in `Stopwatch`.

## Bug Fixes
- Fix incorrect calculation of fraction digits in `NumberFormatter`.

# 2.0.0 (5 Aug 2021)

## Enhancements
- Add percentage information to the output.
- Add `NumberFormatter.percentage()` method to format the given value.
- Add more constructors to `WatchItem`.
- Change some column names in the output to be shorter.
- Support `getTotalElapsedMilliseconds()` and `getTotalElapsedSeconds()` methods in `Stopwatch`.
- Support `add()` method in `Stopwatch` to customize the output.

# 1.0.0 (3 Aug 2021)

## Usage
```java
Stopwatch stopwatch = new Stopwatch();

stopwatch.start("first phase");
// ...
stopwatch.stop();

stopwatch.start("second phase");
// ...
stopwatch.stop();

stopwatch.print();
```

## Result
```
|         name | elapsed(ms) | elapsed(s) |
|--------------|-------------|------------|
|  first phase |     1,234ms |     1.234s |
| second phase |       830ms |     0.830s |
```

src/test/java/org/silentsoft/stopwatch/StopwatchTest.java
package org.silentsoft.stopwatch;

import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.OutputStream;
import java.util.concurrent.atomic.AtomicBoolean;

public class StopwatchTest {

    @Test
    public void stopwatchTest() {
        Stopwatch stopwatch = new Stopwatch();
        stopwatch.start("test-1");
        {
            WatchItem watchItem = stopwatch.watchItems.peekFirst();

            Assertions.assertEquals(1, stopwatch.watchItems.size());
            Assertions.assertNotNull(watchItem);
            Assertions.assertNull(watchItem.getEndEpochMilli());
            Assertions.assertEquals("test-1", watchItem.getName());

            stopwatch.stop("none");

            Assertions.assertEquals(1, stopwatch.watchItems.size());
            Assertions.assertNull(watchItem.getEndEpochMilli());

            stopwatch.stop("test-1");

            Assertions.assertEquals(1, stopwatch.watchItems.size());
            Assertions.assertNotNull(watchItem.getEndEpochMilli());
        }
        stopwatch.start("test-2");
        {
            WatchItem watchItem = stopwatch.watchItems.peekLast();

            Assertions.assertEquals(2, stopwatch.watchItems.size());
            Assertions.assertNotNull(watchItem);
            Assertions.assertNull(watchItem.getEndEpochMilli());
            Assertions.assertEquals("test-2", watchItem.getName());

            stopwatch.stop();

            Assertions.assertEquals(2, stopwatch.watchItems.size());
            Assertions.assertNotNull(watchItem.getEndEpochMilli());
        }
    }

    @Test
    public void runnableTest() {
        Stopwatch stopwatch = new Stopwatch();
        AtomicBoolean launchFlag = new AtomicBoolean(false);
        stopwatch.start("test", () -> {
            launchFlag.set(true);
            try {
                Thread.sleep(1);

                stopwatch.pause();
                Thread.sleep(1);
                stopwatch.resume();
            } catch (InterruptedException e) {

            }
        });
        Assertions.assertTrue(launchFlag.get());
        Assertions.assertEquals(1, stopwatch.watchItems.size());
        Assertions.assertNotNull(stopwatch.watchItems.get(0).getEndEpochMilli());
        Assertions.assertNull(stopwatch.watchItems.get(0).getPauseEpochMilli());
        Assertions.assertNull(stopwatch.watchItems.get(0).getResumeEpochMilli());
        Assertions.assertTrue(stopwatch.watchItems.get(0).getTotalPausedMilli() > 0);
        Assertions.assertTrue(stopwatch.getTotalElapsedSeconds() > 0);
        Assertions.assertTrue(stopwatch.getTotalElapsedMilliseconds() > 0);
    }

    @Test
    public void getTotalElapsedTest() {
        Stopwatch stopwatch = new Stopwatch();
        {
            stopwatch.start("test-1");
            /* stopwatch.stop(); */
        }
        {
            WatchItem watchItem = new WatchItem("test-2", 0, 1000);
            stopwatch.watchItems.add(watchItem);
        }
        {
            WatchItem watchItem = new WatchItem("test-3", 0, 234);
            stopwatch.watchItems.add(watchItem);
        }
        {
            WatchItem watchItem = new WatchItem("test-4", 0, 567);
            watchItem.setTotalPausedMilli(567);
            stopwatch.watchItems.add(watchItem);
        }
        {
            WatchItem watchItem = new WatchItem("test-5", 0, 890);
            watchItem.setTotalPausedMilli(890);
            stopwatch.watchItems.add(watchItem);
        }
        {
            stopwatch.start("test-6");
            /* stopwatch.stop(); */
        }

        Assertions.assertEquals(1234, stopwatch.getTotalElapsedMilliseconds());
        Assertions.assertEquals(1.234, stopwatch.getTotalElapsedSeconds());
    }

    @Test
    public void printTest() {
        {
            Stopwatch stopwatch = new Stopwatch();
            stopwatch.start("test");

            ByteArrayOutputStream outputStream = new ByteArrayOutputStream();
            stopwatch.print(outputStream);

            StringBuilder builder = new StringBuilder();
            builder.append("|  name |    % |  ms |      s |\n");
            builder.append("|-------|------|-----|--------|\n");
            builder.append("|  test |      | N/A |    N/A |\n");
            builder.append("|       |      |     |        |\n");
            builder.append("| total | 100% | 0ms | 0.000s |\n");

            Assertions.assertEquals(builder.toString(), new String(outputStream.toByteArray()));
        }
        {
            Stopwatch stopwatch = new Stopwatch();
            stopwatch.add(new WatchItem("test", 0, 1000));

            ByteArrayOutputStream outputStream = new ByteArrayOutputStream();
            stopwatch.print(outputStream);

            StringBuilder builder = new StringBuilder();
            builder.append("|  name |    % |      ms |      s |\n");
            builder.append("|-------|------|---------|--------|\n");
            builder.append("|  test | 100% | 1,000ms | 1.000s |\n");
            builder.append("|       |      |         |        |\n");
            builder.append("| total | 100% | 1,000ms | 1.000s |\n");

            Assertions.assertEquals(builder.toString(), new String(outputStream.toByteArray()));
        }
        {
            Stopwatch stopwatch = new Stopwatch();
            {
                stopwatch.start("test-1");
                /* stopwatch.stop(); */
            }
            {
                WatchItem watchItem = new WatchItem("test-2", 0, 1);
                stopwatch.watchItems.add(watchItem);
            }
            {
                WatchItem watchItem = new WatchItem("test-3", 0, 10);
                stopwatch.watchItems.add(watchItem);
            }
            {
                WatchItem watchItem = new WatchItem("test-4", 0, 100);
                stopwatch.watchItems.add(watchItem);
            }
            {
                WatchItem watchItem = new WatchItem("test-5", 0, 1000);
                stopwatch.watchItems.add(watchItem);
            }
            {
                WatchItem watchItem = new WatchItem("test-6", 0, 1000);
                watchItem.setTotalPausedMilli(1000);
                stopwatch.watchItems.add(watchItem);
            }
            {
                stopwatch.start("test-7");
                /* stopwatch.stop(); */
            }

            ByteArrayOutputStream outputStream = new ByteArrayOutputStream();
            stopwatch.print(outputStream);

            StringBuilder builder = new StringBuilder();
            builder.append("|   name |     % |      ms |      s |\n");
            builder.append("|--------|-------|---------|--------|\n");
            builder.append("| test-1 |       |     N/A |    N/A |\n");
            builder.append("| test-2 |  0.1% |     1ms | 0.001s |\n");
            builder.append("| test-3 |  0.9% |    10ms | 0.010s |\n");
            builder.append("| test-4 |  9.0% |   100ms | 0.100s |\n");
            builder.append("| test-5 | 90.0% | 1,000ms | 1.000s |\n");
            builder.append("| test-6 |  0.0% |     0ms | 0.000s |\n");
            builder.append("| test-7 |       |     N/A |    N/A |\n");
            builder.append("|        |       |         |        |\n");
            builder.append("|  total |  100% | 1,111ms | 1.111s |\n");

            Assertions.assertEquals(builder.toString(), new String(outputStream.toByteArray()));
        }
    }

    @Test
    public void exceptionTest() {
        Assertions.assertThrows(NullPointerException.class, () -> {
            new Stopwatch().start(null);
        });
        Assertions.assertThrows(NullPointerException.class, () -> {
            new Stopwatch().print(null);
        });
        Assertions.assertThrows(RuntimeException.class, () -> {
            Stopwatch stopwatch = new Stopwatch();
            stopwatch.start("");
            stopwatch.stop("");
            stopwatch.print(new OutputStream() {
                @Override
                public void write(int b) throws IOException {
                    throw new IOException();
                }
            });
        });
        Assertions.assertDoesNotThrow(() -> {
            new Stopwatch().stop();
            new Stopwatch().stop("");
            new Stopwatch().pause();
            new Stopwatch().pause("");
            new Stopwatch().resume();
            new Stopwatch().resume("");
            new Stopwatch().print();
            {
                Stopwatch stopwatch = new Stopwatch();
                stopwatch.start("test");
                stopwatch.start("test");
                stopwatch.stop();
                stopwatch.stop();
                stopwatch.stop();
            }
            {
                Stopwatch stopwatch = new Stopwatch();
                stopwatch.start("test");
                stopwatch.pause("test");
                stopwatch.pause("test");
            }
            {
                Stopwatch stopwatch = new Stopwatch();
                stopwatch.start("test");
                stopwatch.pause("test-1");
            }
            {
                Stopwatch stopwatch = new Stopwatch();
                stopwatch.start("test");
                stopwatch.pause("test");
                stopwatch.resume("test");
                stopwatch.resume("test");
            }
            {
                Stopwatch stopwatch = new Stopwatch();
                stopwatch.start("test");
                stopwatch.resume("test-1");
            }
        });
    }

}


src/test/java/org/silentsoft/stopwatch/WatchItemTest.java
package org.silentsoft.stopwatch;

import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;

import java.lang.reflect.Field;

public class WatchItemTest {

    @Test
    public void constructorTest() {
        WatchItem watchItem = new WatchItem("test", 1000, 3000);
        Assertions.assertEquals("test", watchItem.getName());
        Assertions.assertEquals(1000, watchItem.getStartEpochMilli());
        Assertions.assertEquals(3000, watchItem.getEndEpochMilli());
        Assertions.assertNull(watchItem.getPauseEpochMilli());
        Assertions.assertNull(watchItem.getResumeEpochMilli());
        Assertions.assertEquals(0, watchItem.getTotalPausedMilli());
    }

    @Test
    public void getElapsedMilliTest() {
        WatchItem watchItem = new WatchItem("test", 1000, 3000);
        Assertions.assertEquals(2000, watchItem.getElapsedMilli());
    }

    @Test
    public void stopTest() throws Exception {
        {
            WatchItem watchItem = new WatchItem("test");
            pause(watchItem);
            watchItem.resume();
            watchItem.stop();
            Assertions.assertTrue(watchItem.getTotalPausedMilli() > 0);
            Assertions.assertTrue(watchItem.getEndEpochMilli() > 0);
        }
        {
            WatchItem watchItem = new WatchItem("test");
            pause(watchItem);
            watchItem.stop();
            Assertions.assertTrue(watchItem.getTotalPausedMilli() > 0);
            Assertions.assertTrue(watchItem.getEndEpochMilli() > 0);
        }
    }

    private void pause(WatchItem watchItem) throws Exception {
        Field pauseEpochMilliField = watchItem.getClass().getDeclaredField("pauseEpochMilli");
        pauseEpochMilliField.setAccessible(true);
        pauseEpochMilliField.set(watchItem, 1000L);
    }

}


src/test/java/org/silentsoft/stopwatch/NumberFormatterTest.java
package org.silentsoft.stopwatch;

import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;

public class NumberFormatterTest {

    @Test
    public void percentageTest() {
        Assertions.assertEquals("0.0%", NumberFormatter.percentage(0));
        Assertions.assertEquals("49.1%", NumberFormatter.percentage(49.1));
        Assertions.assertEquals("50.0%", NumberFormatter.percentage(50));
        Assertions.assertEquals("51.9%", NumberFormatter.percentage(51.9));
        Assertions.assertEquals("100%", NumberFormatter.percentage(100));
    }

    @Test
    public void millisecondsTest() {
        Assertions.assertEquals("0ms", NumberFormatter.milliseconds(0));
        Assertions.assertEquals("100ms", NumberFormatter.milliseconds(100));
    }

    @Test
    public void secondsTest() {
        Assertions.assertEquals("0.000s", NumberFormatter.seconds(0));
        Assertions.assertEquals("0.830s", NumberFormatter.seconds(0.83));
        Assertions.assertEquals("1.000s", NumberFormatter.seconds(1));
    }

}


src/main/java/org/silentsoft/stopwatch/NumberFormatter.java
package org.silentsoft.stopwatch;

import java.text.NumberFormat;

/**
 * This class is used to formatting percentage, milliseconds and seconds into human-readable format.</p>
 * <p>Examples:</p>
 * <ul>
 *   <li>{@code NumberFormatter.percentage(0)} returns {@code 0.0%}</li>
 *   <li>{@code NumberFormatter.percentage(100)} returns {@code 100%}</li>
 *   <li>{@code NumberFormatter.milliseconds(100)} returns {@code 100ms}</li>
 *   <li>{@code NumberFormatter.seconds(0)} returns {@code 0.000s}</li>
 *   <li>{@code NumberFormatter.seconds(0.83)} returns {@code 0.830s}</li>
 * </ul>
 * 
 * @see #percentage(double)
 * @see #milliseconds(double)
 * @see #seconds(double) 
 */
public class NumberFormatter {

    private NumberFormatter() { }

    private static NumberFormat percentageFormat;
    private static NumberFormat millisecondsFormat;
    private static NumberFormat secondsFormat;

    /**
     * Returns the given {@code number} as a human-readable string with a percentage sign.</p>
     * Note that the minimum fraction digits and maximum fraction digits are always set to {@code 1} except for {@code 100%}.
     * <p>Examples:</p>
     * <ul>
     *   <li>{@code NumberFormatter.percentage(0)} returns {@code 0.0%}</li>
     *   <li>{@code NumberFormatter.percentage(100)} returns {@code 100%}</li>
     * </ul>
     *
     * @param number the number to be formatted
     * @return a formatted string
     */
    public static String percentage(double number) {
        if (percentageFormat == null) {
            percentageFormat = NumberFormat.getInstance();
            percentageFormat.setMinimumFractionDigits(1);
            percentageFormat.setMaximumFractionDigits(1);
        }

        return (number == 100.0 ? "100" : percentageFormat.format(number)).concat("%");
    }

    /**
     * Returns the given {@code number} as a human-readable string with a {@code ms} suffix.</p>
     * <p>Examples:</p>
     * <ul>
     *   <li>{@code NumberFormatter.milliseconds(0)} returns {@code 0ms}</li>
     *   <li>{@code NumberFormatter.milliseconds(100)} returns {@code 100ms}</li>
     * </ul>
     *
     * @param number the number to be formatted
     * @return a formatted string
     */
    public static String milliseconds(double number) {
        if (millisecondsFormat == null) {
            millisecondsFormat = NumberFormat.getInstance();
        }

        return millisecondsFormat.format(number).concat("ms");
    }

    /**
     * Returns the given {@code number} as a human-readable string with a {@code s} suffix.</p>
     * Note that the minimum fraction digits and maximum fraction digits are always set to {@code 3}.
     * <p>Examples:</p>
     * <ul>
     *   <li>{@code NumberFormatter.seconds(0)} returns {@code 0.000s}</li>
     *   <li>{@code NumberFormatter.seconds(0.83)} returns {@code 0.830s}</li>
     * </ul>
     *
     * @param number the number to be formatted
     * @return a formatted string
     */
    public static String seconds(double number) {
        if (secondsFormat == null) {
            secondsFormat = NumberFormat.getInstance();
            secondsFormat.setMinimumFractionDigits(3);
            secondsFormat.setMaximumFractionDigits(3);
        }

        return secondsFormat.format(number).concat("s");
    }

}


src/main/java/org/silentsoft/stopwatch/Stopwatch.java
package org.silentsoft.stopwatch;

import java.io.OutputStream;
import java.io.OutputStreamWriter;
import java.util.LinkedList;
import java.util.stream.IntStream;
import java.util.stream.LongStream;

/**
 * This class is used to measure the time of execution between the {@code start} and {@code stop} methods.
 * Also, it can be paused and resumed at any time so that exclusive time can be measured.</p>
 * The example below shows how to use this stopwatch to measure the time.
 * <pre>
 *   Stopwatch stopwatch = new Stopwatch();
 *
 *   stopwatch.start("initialization");
 *   // ...
 *   stopwatch.stop();
 *
 *   stopwatch.start("processing");
 *   // ...
 *   stopwatch.stop();
 *
 *   stopwatch.start("rendering");
 *   // ...
 *   stopwatch.stop();
 *
 *   stopwatch.print();
 * </pre>
 * and the output is as follows:
 * <pre>
 *   |           name |     % |      ms |      s |
 *   |----------------|-------|---------|--------|
 *   | initialization | 59.5% | 1,234ms | 1.234s |
 *   |     processing | 40.0% |   830ms | 0.830s |
 *   |      rendering |  0.5% |    10ms | 0.010s |
 *   |                |       |         |        |
 *   |          total |  100% | 2,074ms | 2.074s |
 * </pre>
 *
 * @see #start(String)
 * @see #start(String, Runnable)
 * @see #pause()
 * @see #pause(String)
 * @see #resume()
 * @see #resume(String)
 * @see #stop()
 * @see #stop(String)
 * @see #print()
 * @see #print(OutputStream) 
 */
public class Stopwatch {

    protected static final String NOT_AVAILABLE = "N/A";

    protected LinkedList<WatchItem> watchItems = new LinkedList<>();

    /**
     * Starts measuring time immediately with the given {@code name}.
     *
     * @param name the name of the task to start
     * @throws NullPointerException if the name is {@code null}
     * @see #start(String, Runnable)
     */
    public void start(String name) throws NullPointerException {
        add(new WatchItem(name));
    }

    /**
     * Starts measuring time immediately with the given {@code name} and {@code runnable}.</p>
     * Note that after execute the {@code runnable}, the task will be considered as finished.
     *
     * @param name the name of the task to start
     * @param runnable the runnable used to measure the elapsed time
     * @throws NullPointerException if the name is {@code null}
     * @see #start(String)
     */
    public void start(String name, Runnable runnable) throws NullPointerException {
        WatchItem watchItem = new WatchItem(name);
        add(watchItem);

        runnable.run();

        watchItem.stop();
    }

    /**
     * Adds the given {@code watchItem} to the inner list directly.
     *
     * @param watchItem the watch item containing elapsed time information
     * @throws NullPointerException if the name of the watch item is {@code null}
     */
    public void add(WatchItem watchItem) throws NullPointerException {
        if (watchItem.getName() == null) {
            throw new NullPointerException("Name cannot be null.");
        }

        synchronized (watchItems) {
            watchItems.add(watchItem);
        }
    }

    /**
     * Pauses the last started task.
     *
     * @see #pause(String)
     */
    public void pause() {
        pause(null);
    }

    /**
     * Pauses the last started task matching the given {@code name}. If the {@code name} is {@code null}, the last started task is paused.</p>
     * Note that this method does not pause all tasks started with the same name. It only pauses the last started task.
     *
     * @param name the name of the task to pause
     * @see #pause()
     */
    public void pause(String name) {
        synchronized (watchItems) {
            for (int i= watchItems.size()-1; i>=0; i--) {
                WatchItem watchItem = watchItems.get(i);
                if (name == null || watchItem.getName().equals(name)) {
                    if (watchItem.pause()) {
                        break;
                    }
                }
            }
        }
    }

    /**
     * Resumes the last paused task.
     *
     * @see #resume(String)
     */
    public void resume() {
        resume(null);
    }

    /**
     * Resumes the last paused task matching the given {@code name}. If the {@code name} is {@code null}, the last paused task is resumed.</p>
     * Note that this method does not resume all tasks paused with the same name. It only resumes the last paused task.
     *
     * @param name the name of the task to resume
     * @see #resume()
     */
    public void resume(String name) {
        synchronized (watchItems) {
            for (int i = watchItems.size() - 1; i >= 0; i--) {
                WatchItem watchItem = watchItems.get(i);
                if (name == null || watchItem.getName().equals(name)) {
                    if (watchItem.resume()) {
                        break;
                    }
                }
            }
        }
    }

    /**
     * Stops the last started task.
     *
     * @see #stop(String)
     */
    public void stop() {
        stop(null);
    }

    /**
     * Stops the last started task matching the given {@code name}. If the {@code name} is {@code null}, the last started task is stopped.</p>
     * Note that this method does not stop all tasks started with the same name. It only stops the last started task.
     *
     * @param name the name of the task to stop
     * @see #stop()
     */
    public void stop(String name) {
        synchronized (watchItems) {
            for (int i= watchItems.size()-1; i>=0; i--) {
                WatchItem watchItem = watchItems.get(i);
                if (name == null || watchItem.getName().equals(name)) {
                    if (watchItem.stop()) {
                        break;
                    }
                }
            }
        }
    }

    /**
     * Returns the total elapsed time of all tasks in this stopwatch in milliseconds excluding any paused time.
     *
     * @return the total elapsed time in milliseconds
     * @see #getTotalElapsedSeconds()
     */
    public long getTotalElapsedMilliseconds() {
        Long totalElapsedMilliseconds;
        synchronized (watchItems) {
            totalElapsedMilliseconds = watchItems.stream().flatMapToLong(watchItem -> LongStream.of(watchItem.getElapsedMilli() != null ? watchItem.getElapsedMilli() : 0)).sum();
        }
        return totalElapsedMilliseconds;
    }

    /**
     * Returns the total elapsed time of all tasks in this stopwatch in seconds excluding any paused time.
     *
     * @return the total elapsed time in seconds
     * @see #getTotalElapsedMilliseconds()
     */
    public double getTotalElapsedSeconds() {
        return getTotalElapsedMilliseconds() / 1000.0;
    }

    /**
     * Prints the result of the stopwatch as a table to the default output stream.</p>
     * the following table format is used:
     * <pre>
     *   |           name |     % |      ms |      s |
     *   |----------------|-------|---------|--------|
     *   | initialization | 59.5% | 1,234ms | 1.234s |
     *   |     processing | 40.0% |   830ms | 0.830s |
     *   |      rendering |  0.5% |    10ms | 0.010s |
     *   |                |       |         |        |
     *   |          total |  100% | 2,074ms | 2.074s |
     * </pre>
     * Note that all tasks are cleared after printing.
     *
     * @see #print(OutputStream)
     * @see System#out
     */
    public void print() {
        print(System.out);
    }

    /**
     * Prints the result of the stopwatch as a table to the given output stream.</p>
     * the following table format is used:
     * <pre>
     *   |           name |     % |      ms |      s |
     *   |----------------|-------|---------|--------|
     *   | initialization | 59.5% | 1,234ms | 1.234s |
     *   |     processing | 40.0% |   830ms | 0.830s |
     *   |      rendering |  0.5% |    10ms | 0.010s |
     *   |                |       |         |        |
     *   |          total |  100% | 2,074ms | 2.074s |
     * </pre>
     * Note that all tasks are cleared after printing.
     *
     * @param outputStream the output stream to print
     * @throws NullPointerException if the output stream is null
     * @throws RuntimeException if the output stream cannot be written
     */
    public void print(OutputStream outputStream) {
        if (outputStream == null) {
            throw new NullPointerException("Output stream cannot be null.");
        }

        synchronized (watchItems) {
            if (watchItems.isEmpty()) {
                return;
            }

            try (OutputStreamWriter writer = new OutputStreamWriter(outputStream)) {
                int maxLengthOfName = watchItems.stream().flatMapToInt(watchItem -> IntStream.of(watchItem.getName().length())).max().getAsInt();
                maxLengthOfName = Math.max(maxLengthOfName, Math.max("name".length(), "total".length()));

                long totalElapsedMilliseconds = getTotalElapsedMilliseconds();

                int maxLengthOfPercentage = watchItems.stream().flatMapToInt(watchItem -> IntStream.of(watchItem.getElapsedMilli() == null ? 0 : NumberFormatter.percentage((watchItem.getElapsedMilli() / (totalElapsedMilliseconds * 1.0)) * 100).length())).max().getAsInt();
                maxLengthOfPercentage = Math.max(maxLengthOfPercentage, "100%".length());

                int maxLengthOfElapsedMilliseconds = watchItems.stream().flatMapToInt(watchItem -> IntStream.of(watchItem.getElapsedMilli() == null ? 0 : NumberFormatter.milliseconds(watchItem.getElapsedMilli()).length())).max().getAsInt();
                maxLengthOfElapsedMilliseconds = Math.max(maxLengthOfElapsedMilliseconds, Math.max(NOT_AVAILABLE.length(), NumberFormatter.milliseconds(totalElapsedMilliseconds).length()));

                double totalElapsedSeconds = getTotalElapsedSeconds();
                int maxLengthOfElapsedSeconds = watchItems.stream().flatMapToInt(watchItem -> IntStream.of(watchItem.getElapsedMilli() == null ? 0 : NumberFormatter.seconds(watchItem.getElapsedMilli() / 1000.0).length())).max().getAsInt();
                maxLengthOfElapsedSeconds = Math.max(maxLengthOfElapsedSeconds, Math.max(NOT_AVAILABLE.length(), NumberFormatter.seconds(totalElapsedSeconds).length()));

                String title = String.format("| %s | %s | %s | %s |\n", fillWithWhitespace("name", maxLengthOfName), fillWithWhitespace("%", maxLengthOfPercentage), fillWithWhitespace("ms", maxLengthOfElapsedMilliseconds), fillWithWhitespace("s", maxLengthOfElapsedSeconds));
                writer.write(title);

                String separator = String.format("%s\n", createSeparator(maxLengthOfName, maxLengthOfPercentage, maxLengthOfElapsedMilliseconds, maxLengthOfElapsedSeconds));
                writer.write(separator);

                while (watchItems.isEmpty() == false) {
                    WatchItem watchItem = watchItems.poll();
                    String name = watchItem.getName();

                    String percentage;
                    String elapsedMilliseconds;
                    String elapsedSeconds;

                    Long elapsedMilli = watchItem.getElapsedMilli();
                    if (elapsedMilli == null) {
                        percentage = "";
                        elapsedMilliseconds = NOT_AVAILABLE;
                        elapsedSeconds = NOT_AVAILABLE;
                    } else {
                        percentage = NumberFormatter.percentage((elapsedMilli / (totalElapsedMilliseconds * 1.0)) * 100);
                        elapsedMilliseconds = NumberFormatter.milliseconds(elapsedMilli);
                        elapsedSeconds = NumberFormatter.seconds(elapsedMilli / 1000.0);
                    }

                    writer.write(String.format("| %s | %s | %s | %s |\n", fillWithWhitespace(name, maxLengthOfName), fillWithWhitespace(percentage, maxLengthOfPercentage), fillWithWhitespace(elapsedMilliseconds, maxLengthOfElapsedMilliseconds), fillWithWhitespace(elapsedSeconds, maxLengthOfElapsedSeconds)));
                }

                String emptyRow = String.format("%s\n", createEmptyRow(maxLengthOfName, maxLengthOfPercentage, maxLengthOfElapsedMilliseconds, maxLengthOfElapsedSeconds));
                writer.write(emptyRow);

                String total = String.format("| %s | %s | %s | %s |\n", fillWithWhitespace("total", maxLengthOfName), fillWithWhitespace("100%", maxLengthOfPercentage), fillWithWhitespace(NumberFormatter.milliseconds(totalElapsedMilliseconds), maxLengthOfElapsedMilliseconds), fillWithWhitespace(NumberFormatter.seconds(totalElapsedSeconds), maxLengthOfElapsedSeconds));
                writer.write(total);

                writer.flush();
            } catch (Exception e) {
                throw new RuntimeException(e);
            }
        }
    }

    /**
     * Returns the value with whitespace filled in to the left
     *
     * @param value the value to fill with whitespace
     * @param max the maximum length of the value
     * @return a string representation of the value with whitespace filled in to the left
     */
    protected String fillWithWhitespace(String value, int max) {
        int count = max - value.length();
        for (int i=0; i<count; i++) {
            value = " ".concat(value);
        }
        return value;
    }

    /**
     * Returns separator with the dashes in each column.
     *
     * @param lengthOfColumns the length of each column
     * @return a string which is used to separator for the table header
     */
    protected String createSeparator(int... lengthOfColumns) {
        String separator = "|-";
        for (int i=0, j=lengthOfColumns.length; i<j; i++) {
            int lengthOfColumn = lengthOfColumns[i];
            for (int k=0; k<lengthOfColumn; k++) {
                separator = separator.concat("-");
            }
            if (i < j-1) {
                separator = separator.concat("-|-");
            }
        }
        separator = separator.concat("-|");

        return separator;
    }

    /**
     * Returns empty row with the whitespaces in each column.
     *
     * @param lengthOfColumns the length of each column
     * @return a string which is used to create an empty row
     */
    protected String createEmptyRow(int... lengthOfColumns) {
        return createSeparator(lengthOfColumns).replaceAll("-", " ");
    }

}
